streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -203,13 +202,27 @@ public void registerStore(final StateStore store,
         );
 
         try {
-            restoreState(
-                stateRestoreCallback,
-                topicPartitions,
-                highWatermarks,
-                store.name(),
-                converterForStore(store)
-            );
+            if (topology.storeNameToReprocessOnRestore().getOrDefault(store.name(), false)) {
+                globalConsumer.assign(topicPartitions);
+                globalConsumer.seekToBeginning(topicPartitions);
+                for (final TopicPartition topicPartition : topicPartitions) {
+                    stateRestoreListener.onRestoreStart(topicPartition, store.name(),
+                        checkpointFileCache.getOrDefault(topicPartition, 0L),
+                        checkpointFileCache.getOrDefault(topicPartition, 0L));
+                    stateRestoreListener.onRestoreEnd(topicPartition, store.name(), 0L);
```

Comment:
Where does the actual restore happen?

Note that the original `restoreState()` is the "bootstrapping phase" before we move to `RUNNING`, and we should preserve this behavior. It seem, your PR basically skips the bootstrapping, and relies on the regular processing to re-read the data? For this case, we would go to `RUNNING` with an empty global store and thus lookups might fail as the data is not loaded yet.